blog/2023-02-20-preventing-vulnerabilities-in-solidity-reentrancy-attack/preventing-vulnerabilities-in-solidity-reentrancy-attack.md

```diff
@@ -0,0 +1,350 @@
+---
+
+![header](../../src/data-tutorials/showcase/intermediate/preventing-vulnerabilities-in-solidity-reentrancy-attack.png)
+
+## Introduction
+
+To continue our previous articles on the vulnerabilities of Solidity and how to prevent them, we will dive into the reentrancy attack. We will look at code examples and devise solutions to avoid reentrancy.
+
+## Prerequisites
+
+To properly understand this guide, you must have a solid knowledge of creating smart contracts with Solidity. The link below will come in handy:
+
+https://docs.celo.org/blog/tutorials/solidity-from-zero-to-hero
+
+## Requirements
+
+To code along with this guide, you must set up a text editor or an IDE like [Remix IDE](https://remix.ethereum.org).
+
+## What Is Reentrancy Attack?
+
+In Solidity, reentrancy describes a scenario in which a contract calls a second contract. The second contract then calls the first contract while the first contract is still running. 
+
+For better comprehension, reentrancy means that contract A can call contract B, but while contract A is still executing the call, contract B calls contract A. This behavior will cause the program to go into an infinite loop, ultimately making the contract run out of [gas](https://solidity-by-example.org/gas/).
+
+An attacker can use reentrancy to perform various malicious actions, such as transferring all of the funds in a contract to the attacker's account or modifying the contract's state in unexpected ways.
+
+For better understanding, let us cite an example:
+
+1. `Contract A` has 20 ETH in it. It also keeps a record of how many ETH it owes other contracts. Right now, it owes `contract B` 1 ETH.
+2. `Contract B` has zero ETH right now.
+3. `Contract A` has a function called `payDebt()` inside it. This function will check that the caller's balance exceeds zero, then send the amount owed. Finally, it will reset the caller's balance to zero. This way, it can repay `contract B` and still have 19 ETH left.
+4. `Contract B` has two functions in it. The functions are `attack()` and `fallback()`. Both functions call the `A.payDebt()` function.
+5. First, `contract B` calls `B.attack()`. The call to `B.attack()` will automatically call `A.payDebt()`.
+6. `A.payDebt()` executes sequentially and sends the owed balance before resetting it. When it sends the balance initially, it triggers the `fallback()` function inside `contract B` which then calls the `payDebt()` function again.
+7. Since `contract A` had yet not reset the balance, `contract A` still has a record that it owes `contract B` 1 ETH. Thus, it will keep sending ETH to `contract B`.
+8. `Contract A` continually sends ETH because the function sends the owed amount before resetting it to zero. And the amount is never reset because sending the owed amount triggers another call to the `payDebt()` function. Therefore, for every payment made to `contract B`, `contract A` gets called to make another payment.
+
```

Comment:
Hey just add something like I am Using remix.ide

Reply:
Hello @Ishanpathak1, the tutorial is for intermediate programmers. The tools have been highlighted in the prerequisite.